applications/tari_base_node/src/builder.rs

```diff
@@ -92,83 +89,122 @@ use tari_wallet::{
         storage::sqlite_db::OutputManagerSqliteDatabase,
         OutputManagerServiceInitializer,
     },
-    storage::connection_manager::run_migration_and_create_connection_pool,
+    storage::connection_manager::{run_migration_and_create_connection_pool, WalletConnection},
     transaction_service::{
         config::TransactionServiceConfig,
         handle::TransactionServiceHandle,
         storage::sqlite_db::TransactionServiceSqliteDatabase,
         TransactionServiceInitializer,
     },
 };
-use tokio::runtime::Runtime;
+use tokio::{runtime, stream::StreamExt};
 
 const LOG_TARGET: &str = "base_node::initialization";
 
-pub struct BaseNodeContext {
-    pub wallet_transaction_service: TransactionServiceHandle,
-    pub wallet_output_service: OutputManagerHandle,
-    pub node_service: LocalNodeCommsInterface,
+#[macro_export]
+macro_rules! using_backend {
+    ($self:expr, $i: ident, $cmd: expr) => {
+        match $self {
+            NodeContainer::LMDB($i) => $cmd,
+            NodeContainer::Memory($i) => $cmd,
+        }
+    };
 }
 
-pub enum NodeType {
-    LMDB(BaseNodeStateMachine<LMDBDatabase<HashDigest>>),
-    Memory(BaseNodeStateMachine<MemoryDatabase<HashDigest>>),
+/// The type of DB is configured dynamically in the config file, but the state machine struct has static dispatch;
+/// and so we have to use an enum wrapper to hold the various acceptable types.
+pub enum NodeContainer {
+    LMDB(BaseNodeContext<LMDBDatabase<HashDigest>>),
+    Memory(BaseNodeContext<MemoryDatabase<HashDigest>>),
 }
 
-impl NodeType {
-    pub fn get_flag(&self) -> Arc<AtomicBool> {
-        match self {
-            NodeType::LMDB(n) => n.get_interrupt_flag(),
-            NodeType::Memory(n) => n.get_interrupt_flag(),
-        }
+impl NodeContainer {
+    /// Starts the node container. This entails starting the miner and wallet (if `mining_enabled` is true) and then
+    /// starting the base node state machine. This call consumes the NodeContainer instance.
+    pub async fn run(self, rt: runtime::Handle) {
+        async move { using_backend!(self, ctx, NodeContainer::run_impl(ctx, rt).await) }.await
```

Comment:
I don't think this nested async block needed

Reply:
👍  Removed